### PR TITLE
fix(frontend): recover chat stop UI when cancel ack is lost during reconnect

### PR DESCRIPTION
**Pull request by jnhu76**
## 概要

修复 #1286。

这个 PR 修复一个 chat stop / cancel recovery 场景下的前端状态残留问题：

当用户点击 **Stop generating** 后，前端已经成功发送 `chat:cancel`，但如果 cancel ack 在底层 Engine.IO transport 断开 / Socket.IO 重连窗口中丢失，前端可能会一直停留在 stopping / streaming 状态，导致按钮永久转圈，或者 recovery 后仍残留 `Stop generating` / `Queue message` / `cancel-task-button`。

这个问题不是普通 Stop 路径的问题。正常网络条件下，cancel ack 能返回，main 分支也可以正常恢复。真正出问题的是：

1. 用户点击 **Stop generating**；
2. 前端成功发送 `chat:cancel`；
3. 底层 transport 在 cancel ack 返回前断开；
4. cancel ack 丢失；
5. Socket.IO 自动重连；
6. recovery `task:join` 返回服务端已经没有 active stream；
7. 但前端仍被 stale stopping / streaming / running 状态拉住，无法恢复到 Send。

---

## 根因

最终确认有几层问题：

1. `cancelChatStream` 只依赖 Socket.IO ack callback。ack 丢失时，Promise 可能永久 pending。
2. `messageSyncer.stopStream` 会卡在 `await cancelChatStream(...)`，导致 stopping 状态不能及时清理。
3. recovery `task:join` 已经返回 no streaming，但本地 stale streaming message / stale `taskStatus='RUNNING'` 仍可能把 UI 派生为 `isStreaming=true`。
4. recovery no-stream 后，task-level 派生状态仍可能暴露 stale `cancel-task-button` / queue 行为。

---

### 1. `SocketContext.tsx`

- 给 `cancelChatStream` 增加 10 秒 ack timeout。
- cancel ack 丢失时不再永久 pending。
- 增加 settled guard，避免 ack 和 timeout 双重 resolve。

### 2. `messageSyncer.tsx`

- 确保 stop 流程在 timeout / recovery 路径后清理 stopping 状态。
- 将 cancel ack timeout 视为可恢复事件，而不是让 UI 永久停留在 stopping。
- timeout 日志改为非阻塞 warning 语义，避免误导为不可恢复错误。

### 3. `TaskStateMachine.ts`

- 增加 `serverConfirmedNoStream`。
- 当 recovery `task:join` 返回 no streaming，并且子任务处于 terminal 状态时，设置 `serverConfirmedNoStream=true`。
- 当服务端确认 no stream 后，抑制 stale `canCancelTask` / `blocksQueuedDispatch`。
- 当本地仍有 stale streaming message，但服务端返回 no stream 且子任务 terminal 时，将 stale streaming message finalize。
- 在 `SYNC_DONE_STREAMING`、`SEND_ACCEPTED`、terminal lifecycle handling 中清除 `serverConfirmedNoStream`，确保下一轮 streaming 不被污染。

### 4. `useChatStreamHandlers.tsx`

修正 UI 派生逻辑，避免 stale `taskStatus='RUNNING'` 在 server confirmed no-stream 后强制 `isStreaming=true`。

核心逻辑：

```ts
isStreaming =
  isMachineStreaming ||
  (isRunningLifecycle && !serverConfirmedNoStream) ||
  hasPendingUserMessage
```

## A/B 验证说明

验证阶段临时加入过 source-level fault injection。该实验代码已经从最终 PR 中删除。

这里的实验不是简单“永久断开 WebSocket”，而是模拟真实的 cancel ack 丢失窗口：

1. 正常发送 `chat:cancel`；
2. 发送后关闭底层 `Engine.IO transport`；
3. cancel ack 在 transport 断开 / 重连窗口中丢失；
4. Socket.IO 自动 reconnect；
5. reconnect 后走真实 `task:join` recovery；
6. recovery 返回 no streaming；
7. 前端根据 server confirmation 修正 UI。

也就是说，测试的是：

> `chat:cancel` 已经成功发出，但 ack 丢失；Socket.IO 自动重连后，客户端依赖 recovery join ack 恢复正确 UI 状态。

这更接近真实网络波动 / 连接切换场景，而不是人为让客户端永久 offline。

## A/B 测试结果

| 场景	| main | 本 PR |
| --- | --- | --- |
| 正常 Stop，socket 连接正常 | ✅ 正常恢复 | ✅ 正常恢复 |
| 正常 streaming 下 Stop + Queue 双按钮	| ✅ baseline 行为 | ✅ 未改变 |
| cancel ack 丢失 + transport 断开 + 自动重连 | ❌ stuck stopping spinner |  ✅ 恢复到 Send |
| recovery 后发送新消息 | — | ✅ Send，不是 Queue |
| recovery 后再次 Stop | — | ✅ 正常恢复 |

## PR 分支故障注入验证链路 

验证时观察到的关键链路：

```bash
chat:cancel sent
Engine.IO transport closed after cancel send
Socket.IO reconnect attempt
Socket.IO reconnected
task:join ack (no streaming)
TaskStateMachine recover join ack
SYNC_DONE(server confirmed no stream)
cancel acknowledgement timed out
```

这里的 timeout 是预期结果：它只表示本地没有收到 cancel ack。最终状态以后续 recovery `task:join ack(no streaming)` 为准。

## UI 验收

recovery 完成后已验证：

- ✅ 没有 stuck stopping spinner；
- ✅ 没有 stale `Stop generating`；
- ✅ 没有 stale `Queue message`；
- ✅ 没有 `cancel-task-button`；
- ✅ 空输入时是 disabled Send；
- ✅ 有输入时是 enabled Send；
- ✅ 没有发出 `POST /api/tasks/:id/cancel`；
- ✅ recovery 后输入新文本，按钮是 Send，不是 Queue；
- ✅ 新一轮 Send / streaming / Stop 正常。

## baseline 行为保持不变

正常 active streaming 期间，main 分支本来就可能同时显示：

- `Stop generating`
- `Queue message`

这看起来是已有的 queue-during-streaming 行为，不属于本 PR 的修复范围。

本 PR 修复的是：

> recovery 已经确认 no active stream 后，UI 不应继续残留 Stop / Queue / cancel-task 状态。

## 测试

本地验证：

```bash
npx tsc --noEmit
npx eslint
npx jest --no-coverage
git diff --cached --check
```

结果：

```bash
163 test suites passed
1042 tests passed
1 todo
```

测试覆盖包括：

- cancel ack timeout 不永久 pending；
- normal ack 正常返回；
- timeout 后 late ack 不应二次 resolve；
- recovery `task:join ack(no streaming)` 后 UI 回到 Send；
- stale `taskStatus='RUNNING'` 不应在 server confirmed no-stream 后显示 Stop / Queue / Cancel；
- `CANCELLED` terminal subtask 也能触发 no-stream recovery；
- `SEND_ACCEPTED` 后清除 `serverConfirmedNoStream`；
- 正常 active streaming baseline 不改变。

## 清理说明

验证阶段使用的临时故障注入代码已删除。

最终 PR 不包含：

- `WEGENT_TEST_CANCEL_CLOSE_AFTER_SEND`
- `__TEST_CANCEL_CLOSE_AFTER_SEND`
- `[TEST]` 实验日志
- transport close 实验逻辑
- fault injection 代码

最终 PR 只保留正式修复逻辑和自动化回归测试。

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced chat stream cancellation with timeout enforcement and server state synchronization.
  * Fixed UI streaming state to properly reflect server confirmation of no active stream, preventing stale "streaming" display during interruptions.

* **Tests**
  * Added test coverage for stream cancellation timeout scenarios and recovery edge cases.

### CODE REVIEW

File: frontend/src/__tests__/features/tasks/state/TaskStateMachine.test.ts [hunk 1]
```diff
@@ -1512,4 +1513,509 @@ describe('TaskStateMachine', () => {
...
+      await machine.recover({ force: true })
+      expect(machine.getState().derived.serverConfirmedNoStream).toBe(true)
+
+      machine.markSendAccepted('2026-06-08T19:05:00.000Z')
+
+      expect(machine.getState().derived.serverConfirmedNoStream).toBe(false)
+
+      consoleInfoSpy.mockRestore()
+    })
+  })
```
[the remark this answers is not in the document]
Comment (jnhu76) [reply frontend/src/__tests__/features/tasks/state/TaskStateMachine.test.ts]: Partially addressed. The newly added recovery-focused suites were moved into TaskStateMachine.recovery.test.ts. I avoided broader test refactoring to keep this PR focused.

File: frontend/src/__tests__/features/tasks/state/TaskStateMachine.test.ts [hunk 2]
```diff
@@ -1512,4 +1513,509 @@ describe('TaskStateMachine', () => {
     expect(state.runtime.activeStreamSubtaskId).toBeUndefined()
     expect(state.derived.blocksQueuedDispatch).toBe(false)
   })
+
+  describe('recovery invariant: transient isStopping cleared after sync completes', () => {
+    const consoleInfoSpy = jest.spyOn(console, 'info').mockImplementation(() => {})
+
+    afterAll(() => {
+      consoleInfoSpy.mockRestore()
+    })
```
[the remark this answers is not in the document]
Comment (jnhu76) [reply frontend/src/__tests__/features/tasks/state/TaskStateMachine.test.ts]: Fixed. Moved console.info spy setup/teardown to per-test hooks.

File: frontend/src/features/tasks/state/TaskStateMachine.ts [hunk 3]
```diff
@@ -1358,7 +1369,41 @@ export class TaskStateMachine {
...
+            staleSubtask.status === 'CANCELLED' ||
+            staleSubtask.status === 'FAILED')
+
+        if (isStaleSubtaskTerminal) {
+          const staleMsgId = generateMessageId('ai', staleStreamingSubtaskId)
+          const staleMsg = this.state.messages.get(staleMsgId)
+          if (staleMsg) {
+            const newMessages = new Map(this.state.messages)
+            newMessages.set(staleMsgId, { ...staleMsg, status: 'completed' })
+            this.state = { ...this.state, messages: newMessages }
```
[the remark this answers is not in the document]
Comment (jnhu76) [reply frontend/src/features/tasks/state/TaskStateMachine.ts:1420]: Fixed. Stale stream finalization now preserves terminal subtask semantics:

* `COMPLETED` maps to local message status `completed`
* `FAILED` maps to local message status `error`
* `CANCELLED` maps to local message status `error`

I also added regression coverage for the `FAILED` stale subtask case.

File: frontend/src/contexts/SocketContext.tsx [hunk 4]
```diff
@@ -573,7 +581,11 @@ export function SocketProvider({ children }: { children: ReactNode }) {
             shell_type: shellType,
           },
           (response: { success?: boolean; error?: string }) => {
-            resolve({ success: response.success ?? true, error: response.error })
+            if (!settled) {
+              settled = true
+              clearTimeout(timer)
+              resolve({ success: response.success === true, error: response.error })
+            }
```
[the remark this answers is not in the document]
Comment (jnhu76) [reply frontend/src/contexts/SocketContext.tsx]: Fixed. The cancel ack payload is now optional, and the handler returns early if the promise has already settled. Timeout cleanup and promise resolution now happen only once.

Missing or malformed ack payloads no longer throw, and only an explicit `success: true` is treated as a successful cancellation.

**Comment by jnhu76** [comment]
Converting this PR back to draft temporarily.

The PR has grown beyond the original cancel-ack timeout fix and now touches task runtime recovery semantics. Before requesting final review again, I want to verify and patch the remaining `serverConfirmedNoStream` lifecycle issues:

1. server no-stream fast path should set `serverConfirmedNoStream: true`;
2. chat-driven streaming re-entry should clear it back to `false`.

I’ll keep the follow-up patch minimal and avoid broad refactors such as extracting `doSync()` or general test cleanup.

**Comment by jnhu76** [comment]
I’m closing this draft PR in favor of the cleaner replacement PR: #1320 .

This branch was useful as an investigation branch. It helped identify the full failure matrix around:

* lost `chat:cancel` ack during reconnect;
* server-confirmed no-stream recovery;
* stale RUNNING / streaming UI state;
* stale stream event resurrection;
* stale pending chunks / late `CHAT_CHUNK` / late `CHAT_START`;
* terminal stale stream semantics.

The replacement PR is rebuilt from latest `main`, carries over only the confirmed minimal fixes and focused regression tests, and avoids the broader diff growth from this investigation branch.
